vcap/vcap/batch_executor.py

```diff
@@ -57,33 +58,41 @@ def __init__(self, batch_fn: Callable[[List[Any]], List[Any]],
     def total_imgs_in_pipeline(self) -> int:
         return self._request_queue.qsize() + self._num_imgs_being_processed
 
-    def submit(self, input_data: Any, output_queue: Queue = None) -> Queue:
-        """Creates an OvenRequest for you and returns the output queue"""
-        output_queue = output_queue if output_queue else Queue()
-        self._request_queue.put(_OvenRequest(
-            output_queue=output_queue,
+    def submit(self, input_data: Any, future: Future = None) -> Future:
+        """Submits a job and returns a Future that will be fulfilled later."""
+        future = future or Future()
+
+        # Add backwards compatibility for 0.2
+        future.get = future.result
+        future.get = deprecated(
+            message="Use future.result() in place of future.get()",
+            remove_in="0.3.0"
+        )(future.get)
+
+        self._request_queue.put(_Request(
+            future=future,
             input_data=input_data))
-        return output_queue
+        return future
 
-    def _on_requests_ready(self, batch: List[_OvenRequest]) -> None:
-        """Push images through the given prediction backend
+    def _on_requests_ready(self, batch: List[_Request]) -> None:
+        """Push inputs through the given prediction backend
 
         :param batch: A list of requests to work on
         """
-        # Extract the images from the requests
+        # Extract the futures from the requests
         inputs: List[Any] = [req.input_data for req in batch]
-        output_queues: List[Queue] = [req.output_queue for req in batch]
+        futures: List[Future] = [req.future for req in batch]
 
         # Route the results to each request
-        predictions: List[Any] = self.batch_fn(inputs)
-
-        response_count = 0
-        for response_count, output in enumerate(predictions, start=1):
-            output_queues[response_count - 1].put(output)
-
-        if response_count != len(inputs):
-            logging.error(f"CRITICAL ERROR: Backend returned {response_count} "
-                          f"responses. Expected {len(inputs)}")
+        try:
+            for prediction in self.batch_fn(inputs):
+                # Popping the futures ensures that if an error occurs, only
+                # the futures that haven't had a result set will have
+                # set_exception called
+                futures.pop(0).set_result(prediction)
+        except BaseException as exc:
```

Comment:
Use `Exception` not `BaseException`

Reply:
I'm going to use BaseException, because this is emulating the code from the ThreadPoolExecutor in the standard library. Their workers catch BaseException.

Reply:
I could put a link to the pattern I'm emulating in a comment there perhaps?
https://github.com/python/cpython/blob/91e93794d5dd1aa91fbe142099c2955e0c4c1660/Lib/concurrent/futures/thread.py#L73

The idea being that under no circumstances should the worker fail, but it should pass up all exceptions via futures

Reply:
That sounds good
Note that you can use a short url to make it less ugly in the comment
https://github.com/python/cpython/blob/91e93794/Lib/concurrent/futures/thread.py#L51

Reply:
I like this very much, I shall add it

Reply:
Added